### DISCUSSION TITLE
Directly convert the Zim notebook?

### DISCUSSION THREAD

**Discussion by peter88213** [peter88213/zim2obsidian#8] [Ideas]
Currently (v0.4), *zim2obsidian* relies on Zim's Markdown export. However, it also could convert the original Zim notebook, like [pschwede's script](https://github.com/pschwede/zim2md) does. 

This would require parsing the notebook pages. Perhaps some Zim code could be used for this? The Zim Markdown export is specified [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/formats/markdown.py) and [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/formats/plain.py).

The goal would be to build the data structure of the notebook as an object tree, in order to generate arbitrary new structures (XML would also be conceivable) from it. 

The question is whether the whole thing is not already laid out in the Zim code, and whether one should not better start there, for example with a plugin or a contribution to the core.

**Comment by sojusnik** [comment]
I was also pondering about this, because after trying both scripts I tend to that after converting with [pschwede's script](https://github.com/pschwede/zim2md) there are fewer issues and those seem to be easier to fix from a layman's point of view. But pschwede didn't react to my [request](https://github.com/pschwede/zim2md/issues/2), so I assume he isn't interested in improving his script. When skimming through the source code, I've seen some comments in the code, so he must be aware that some things are still outstanding.

So considering the speed of implementation and compatibility (because we don't know how the built-in export function in Zim will work in the feature) a direct conversion might be preferable.

It would be very helpful to have a definitive decision on this question (direct or indirect conversion) as soon as possible, to effectively debug the conversion.

@peter88213 What do you think?

**Reply by peter88213** [reply]
I guess, @pschwede created his script for himself, and it did the job for him. Some portions of his code look a bit odd to me, and there is no license information (which indicates to me that it isn't meant to be a maintained project), but who cares? After all, it's available on the web, ready to be improved by anyone.

It is necessary to realize that the purpose of this software is one-time use, and the programming effort for it should be minimal. The currently available scripts for Zim data conversion are still small and manageable enough that even less experienced Python programmers have a chance to quickly understand and adapt them for their purposes. 

What I described at the beginning of this thread, on the other hand, would be a full-blown application, a software product, so to speak, that should be maintained by a community of developers. The question is: who has an interest in it? Who wants to migrate large amounts of data on a regular basis, and needs an optimized, smooth workflow to do so? 

You must never forget that the main motive of open source software developers is their own needs. As far as I am concerned, it is met with the *zim2obsidian* script as it is now. I'm actually writing the whole thing here just to encourage successors.

**Reply by sojusnik** [reply]
> It is necessary to realize that the purpose of this software is one-time use, and the programming effort for it should be minimal.

Which would lead us to the direct conversion as it is done in [pschwede's script](https://github.com/pschwede/zim2md) I guess.

> What I described at the beginning of this thread, on the other hand, would be a full-blown application, a software product, so to speak, that should be maintained by a community of developers. The question is: who has an interest in it? Who wants to migrate large amounts of data on a regular basis, and needs an optimized, smooth workflow to do so?

Are you speaking of a universal script that would allow converting between different formats or only between Zim and Obsidian? If the latter, then I would assume that the interest would be minimal, because a conversion is usually a one-time action, so the demand for regularly converting from Zim to Obsidian and vice versa is quite low.

**Reply by peter88213** [reply]
> > It is necessary to realize that the purpose of this software is one-time use, and the programming effort for it should be minimal.
> 
> Which would lead us to the direct conversion as it is done in [pschwede's script](https://github.com/pschwede/zim2md) I guess.

Note that I am talking about programming effort, not user effort. In my opinion, post-processing the Markdown pages exported by Zim is much easier than parsing the Zim-formatted notes.  

Edit:
To make clear that my script is a post-processor rather than a converter, I've changed wording now.

**Reply by sojusnik** [reply]
> Note that I am talking about programming effort, not user effort. In my opinion, post-processing the Markdown pages exported by Zim is much easier than parsing the Zim-formatted notes.

I was also speaking about the programming effort, but because I'm lacking expertise in coding I can't give a reliable estimation which one of those two present approaches would be easier to tweak.

One drawback of the post-processing approach is the lesser degree of control and therefore reliability/compatibility of the migration process, because we don't have full control how the markdown export mechanism in Zim will work in the future. So this post-processing script will only work reliably until some changes are made there.

Also, the markdown export function in Zim has already some flaws as described [here](https://github.com/peter88213/zim2obsidian/issues/11) and [here](https://github.com/peter88213/zim2obsidian/issues/12). I actually don't know where to report them, here, in Zim or further upstream in the Pandoc repository? And I also imagine that even if they are fixed swiftly in the Zim or Pandoc repository, it will take ages before a new Zim release will come out with the implemented fix.

**Reply by peter88213** [reply]
Theoretically, you could modify Zim's Markdown export to drop out the Obsidian style. 
Even more elegant: An additional obsidian export option, if possible, implemented as a plugin. 

Zim is open source, so this change would be possible for an experienced Python programmer without the involvement of the Zim maintainers. In any case, the Zim Python code looks clean and clear to me. However, I'm not going to do that, for reasons I've already explained. 

Now before we rant about Zim's Markdown exporter, we should fairly admit that Zim doesn't use Markdown for a good reason, because Markdown just has its limitations.

**Comment by pschwede** [comment]
> But pschwede didn't react to my https://github.com/pschwede/zim2md/issues/2, so I assume he isn't interested in improving his script.

Well, it's low priority (on obsidian) plus I'm not logged in very often. Feel free to clone the project if you like it. I think a complete [AST parser with support for all](https://pandoc.org/) the different markdown forks and wiki syntax could be worth a startup. I simply gave it a shot with a few regexs to see what my zim-wiki looks like in obsidian and logseq. 

> I would assume that the interest would be minimal, because a conversion is usually a one-time action, so the demand for regularly converting from Zim to Obsidian and vice versa is quite low.

Not sure about that. If both directions were implemented correctly, why not try keeping zim and obsidian etc. synchronized?

**Reply by peter88213** [reply]
Yes, these are interesting ideas. However, for me the topic is done. I made my *zim2obsidian* script initially for my own use, and since all my Zim pages are no more complex in design than Markdown pages would be, it served its purpose well. For general use, the concept of post-processing Zim's Markdown export may be too limited, as formatting that cannot be exported to Markdown is lost. 

Here your approach of direct conversion between Zim and Obsidian formatting is more powerful, and if someone wants to continue in this field, your project would be a good starting point. You could encourage people to fork your repository by clarifying the copyright, e.g. by putting your code under a license.

Cheers, 
Peter

**Reply by peter88213** [reply]
Speaking of *Pandoc*, a [custom reader](https://pandoc.org/custom-readers.html) for the Zim format would be a nice thing. A writer already seems to exist. It looks like you don't actually need much more than knowledge of [Lua](https://www.lua.org/about.html).
